packages/common/lib/functions/DPoP.ts

```diff
@@ -0,0 +1,218 @@
+import { jwtDecode } from 'jwt-decode';
+import SHA from 'sha.js';
+import * as u8a from 'uint8arrays';
+import { v4 as uuidv4 } from 'uuid';
+
+import { JwtHeader, JwtPayload, SigningAlgo } from '../types';
+import { JWK } from '../types/CredentialIssuance.types';
+
+import { calculateJwkThumbprint } from './JwkThumbprint';
+import { CreateJwtCallback, JwtIssuerJwk } from './JwtIssuer';
+import { VerifyJwtCallbackBase } from './JwtVerifier';
+
+
+export interface DPoPJwtIssuerWithContext extends JwtIssuerJwk {
+  type: 'dpop';
+  dPoPSigningAlgValuesSupported?: string[];
+}
+
+/**
+ * Returns the current unix timestamp in seconds.
+ */
+function epochTime() {
+  return Math.floor(Date.now() / 1000);
+}
+
+export type DPoPJwtPayloadProps = {
+  htu: string;
+  iat: number;
+  htm: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'HEAD' | 'OPTIONS' | 'TRACE' | 'CONNECT' | 'PATCH';
+  ath?: string;
+  nonce?: string;
+  jti: string;
+};
+export type DPoPJwtHeaderProps = { typ: 'dpop+jwt'; alg: SigningAlgo; jwk: JWK };
+export type CreateDPoPJwtPayloadProps = Omit<DPoPJwtPayloadProps, 'iat' | 'jti' | 'ath'> & { accessToken?: string };
+
+export interface CreateDPoPOptions<JwtPayloadProps = CreateDPoPJwtPayloadProps> {
+  createJwtCallback: CreateJwtCallback<DPoPJwtIssuerWithContext>;
+  jwtIssuer: Omit<JwtIssuerJwk, 'method' | 'type'>;
+  jwtPayloadProps: Record<string, unknown> & JwtPayloadProps;
+  dPoPSigningAlgValuesSupported?: string[];
+}
+
+export type CreateDPoPClientOptions = CreateDPoPOptions<Omit<CreateDPoPJwtPayloadProps, 'htm' | 'htu'>>;
+
+export async function createDPoP(options: CreateDPoPOptions): Promise<string> {
+  const { createJwtCallback, jwtIssuer, jwtPayloadProps, dPoPSigningAlgValuesSupported } = options;
+
+  if (jwtPayloadProps.accessToken && (jwtPayloadProps.accessToken?.startsWith('DPoP ') || jwtPayloadProps.accessToken?.startsWith('Bearer '))) {
+    throw new Error('expected accessToken without scheme');
```

Comment:
I called it a scheme because the specification calls it a scheme. I personally like that because it's easier to find things in the specifications later. So, I'm not sure. Should I change it?

Reply:
No if the spec calls it like that, let's follow it